### PR TITLE
Add Grid stride pairwise dist and fused L2 NN kernels 

### PR DESCRIPTION
**Pull request by mdoijade**
This PR addresses issues mentioned in https://github.com/rapidsai/raft/issues/221
-- Adds grid stride based fusedL2NN kernel, this gives approx 1.85x speed up over previous version of this kernel.
-- Adds support in pairwise dist base class to work for any input size by adding support for grid stride based work distribution.

### CODE REVIEW

**Comment by mdoijade** [comment]
@teju85 for help with reviewing and adding appropriate label

File: cpp/include/raft/distance/cosine.cuh
Comment (teju85) [inline cpp/include/raft/distance/cosine.cuh:86]: minor nitpick: this can be marked `constexpr`.

File: cpp/include/raft/distance/euclidean.cuh
Comment (teju85) [inline cpp/include/raft/distance/euclidean.cuh:94]: same as before: `constexpr`

File: cpp/include/raft/distance/pairwise_distance_base.cuh
Comment (teju85) [inline cpp/include/raft/distance/pairwise_distance_base.cuh:303]: instead of a constant factor of `2` how about computing this at runtime based on the occupancy of the kernel to be launched?

**Review by teju85 — CHANGES_REQUESTED** [review]

File: cpp/include/raft/distance/pairwise_distance_base.cuh
Comment (mdoijade) [reply cpp/include/raft/distance/pairwise_distance_base.cuh]: done. making use of cudaOccupancyMaxActiveBlocksPerMultiprocessor for this purpose now.

**Comment by mdoijade** [comment]
some intermittent CI related issue can you rerun the tests @teju85 
https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci-v0.20/job/raft/job/prb/job/raft-gpu-test/CUDA=11.2,GPU_LABEL=gpu,OS=ubuntu18.04,PYTHON=3.8/62/console

**Comment by mdoijade** [comment]
@teju85 can we merge this ? if not in branch-21.06 then in branch-21.08?

**Review by teju85 — APPROVED** [review]
Changes LGTM.

**Comment by teju85** [comment]
@BradReesWork and/or @afender can we get approval for this from cugraph side? (I believe since you are not using pairwise distances logic anywhere in cugraph, the changes in this PR should not affect you)

**Comment by teju85** [comment]
@divyegala you are still listed as a code owner and thus we'll need your approval too before merging!

**Review by divyegala — APPROVED** [review]
LGTM

**Comment by mdoijade** [comment]
@teju85 @dantegd can we merge this now? the cuML side test PR is working fine as well the L1 dist intermittent issue seems resolved - https://github.com/rapidsai/cuml/pull/3891

**Review by afender — APPROVED** [review]
Lgtm!.

**Comment by teju85** [comment]
@gpucibot merge

**Pull request merged**
